### ISSUE TITLE
Use standalone OpenSSL installation for Windows testing

### ISSUE THREAD

**Issue by ionutbalutoiu** [openenclave/openenclave#1941]
Currently, we use `openssl` bundled with the Git Windows installation from:
`C:\Program Files\Git\mingw64\bin`

Recently, we installed a newer version of Git on our Jenkins Windows servers, and the `crypto`, `crypto_crls_cert_chains` tests started failing on Windows. This happened because `openssl` bundled with the old Git installation was version `1.0.2`, while the newer Git installation had `openssl 1.1.1`.

The `openssl` versions `1.0.2` and `1.1.1` differ in functionality, and for now our tests only work with `1.0.2`. See https://github.com/microsoft/openenclave/pull/1926#issuecomment-501516179 for more context about `openssl 1.1.1` in our tests.

Therefore, it feels cleaner to have a standalone [OpenSSL Windows](https://slproweb.com/products/Win32OpenSSL.html) installation, and do not depend on the one from Git. This way, we won't get regressions when upgrading Git, and we control easier the `openssl` version installed.

**Comment by anitagov** [comment]
Also wanted to add that from a security perspective, it makes sense to have a separate openssl installation although we are only using this to generate certificates on Windows.

**Comment by anitagov** [comment]
@CodeMonkeyLeet @johnkord

**Comment by CodeMonkeyLeet** [comment]
I'm not convinced that this is necessary or sufficient as a mitigation:

- It seems like the procedural cause of the CI blockage is because the update to the CI/CD environments can happen independent of a code review and build consistency checks.

- For example, the ansible config for windows has specified git v2.20 at least as far back in master branch as fcc6b841. That would have brought in OpenSSL v1.1.1a already which should have broken the build when the crypto tests tried to be enabled on Windows. 

- Even if we decide to track OpenSSL as a standalone install, an attempt to update OpenSSL would have surprised us in the same way since the configuration change independent of code changes entering the build are difficult to correlate.

- I would suggest that the configurations of CI servers are only updated using artifacts checked into the branch (i.e. the ansible playbooks). This also means that updates to the ansible playbooks are tested in CI/CD on merge to master as well. I'm curious as to why this isn't already the case.

Given the ideal that we would like the developer set up scripts, our CI/CD environment and the ACC dev environment to be set up in consistent ways, I would prefer to reduce the number of explicit dependencies that need to be installed separately. i.e. using a standalone OpenSSL install for Windows in CD also means adding that to manual instructions for maintainers & ISV devs, otherwise, we would not be testing the configuration we assert we support.

Ultimately, we do want to support latest Git with default OpenSSL and so we should be testing against it anyway. I would argue that use of an OpenSSL configuration is primarily useful for validating that we support backwards compatibility, but _not_ as a mitigation for #1926.

The fact that the crypto tests fail on Windows reflect an actual bug in the lack of robustness in either the test code or the runtime itself when provided with slightly different inputs. Those need to be addressed independently. The issue in CI is the fact that we did not catch that on PR but introduced it as a manual CI configuration action.

@anitagov What security perspective are you referring to where you expect this separation to be preferable? As far as I understand it, this issues is about CI/CD tooling compatibility and the use of standalone OpenSSL in that environment has neither impact to the security of the build process nor the resulting build output.

**Comment by anitagov** [comment]
> I'm not convinced that this is necessary or sufficient as a mitigation:
> 
> * It seems like the procedural cause of the CI blockage is because the update to the CI/CD environments can happen independent of a code review and build consistency checks.
> * For example, the ansible config for windows has specified git v2.20 at least as far back in master branch as [fcc6b84](https://github.com/microsoft/openenclave/commit/fcc6b84112d352a22c010cba054dfd73c5d9a67f). That would have brought in OpenSSL v1.1.1a already which should have broken the build when the crypto tests tried to be enabled on Windows.
> * Even if we decide to track OpenSSL as a standalone install, an attempt to update OpenSSL would have surprised us in the same way since the configuration change independent of code changes entering the build are difficult to correlate.
> * I would suggest that the configurations of CI servers are only updated using artifacts checked into the branch (i.e. the ansible playbooks). This also means that updates to the ansible playbooks are tested in CI/CD on merge to master as well. I'm curious as to why this isn't already the case.

The ansible playbooks had been checked into the master but were deployed on all or several of the Windows CI machines for sandbox testing resulting in the CI failures. 

> 
> Given the ideal that we would like the developer set up scripts, our CI/CD environment and the ACC dev environment to be set up in consistent ways, I would prefer to reduce the number of explicit dependencies that need to be installed separately. i.e. using a standalone OpenSSL install for Windows in CD also means adding that to manual instructions for maintainers & ISV devs, otherwise, we would not be testing the configuration we assert we support.

There is an outstanding request for a power-shell script to install the tools on Windows as developers cannot run Ansible scripts on Windows.

> 
> Ultimately, we do want to support latest Git with default OpenSSL and so we should be testing against it anyway. I would argue that use of an OpenSSL configuration is primarily useful for validating that we support backwards compatibility, but _not_ as a mitigation for #1926.
> 
> The fact that the crypto tests fail on Windows reflect an actual bug in the lack of robustness in either the test code or the runtime itself when provided with slightly different inputs. Those need to be addressed independently. The issue in CI is the fact that we did not catch that on PR but introduced it as a manual CI configuration action.
> 
> @anitagov What security perspective are you referring to where you expect this separation to be preferable? As far as I understand it, this issues is about CI/CD tooling compatibility and the use of standalone OpenSSL in that environment has neither impact to the security of the build process nor the resulting build output.

We use several Linux tools on Windows with the applications provided by Git - bash, perl, python and openssl, etc. However, since OpenSSL is a cryptography/TLS toolkit, we may need a newer OpenSSL to address a new security issue and Git may not have that.

The following 2 need to be addressed in any case - 
- We definitely need to fix the core/tests to support the new features in OpenSSL (CR/LF introduced with certificates, etc).
- The Getting Started Windows Guide will need to be updated to reflect any changes that are being made to tools.

**Comment by ionutbalutoiu** [comment]
> > I'm not convinced that this is necessary or sufficient as a mitigation:
> > 
> > * It seems like the procedural cause of the CI blockage is because the update to the CI/CD environments can happen independent of a code review and build consistency checks.
> > * For example, the ansible config for windows has specified git v2.20 at least as far back in master branch as [fcc6b84](https://github.com/microsoft/openenclave/commit/fcc6b84112d352a22c010cba054dfd73c5d9a67f). That would have brought in OpenSSL v1.1.1a already which should have broken the build when the crypto tests tried to be enabled on Windows.
> 
> Agree.
> 
> > * Even if we decide to track OpenSSL as a standalone install, an attempt to update OpenSSL would have surprised us in the same way since the configuration change independent of code changes entering the build are difficult to correlate.
> > * I would suggest that the configurations of CI servers are only updated using artifacts checked into the branch (i.e. the ansible playbooks). This also means that updates to the ansible playbooks are tested in CI/CD on merge to master as well. I'm curious as to why this isn't already the case.
> 
> The ansible playbooks had been checked into the master but were deployed on all or several of the Windows CI machines for sandbox testing resulting in the CI failures. Things were backed out but not entirely.

So far, for every Ansible change that we check-in, I try to manually apply them on our CI servers. But, most of the time, I apply the changes atomically. For example: If we add new Ansible tasks, I'd make sure I run only those new tasks, instead of all the Ansible setup scripts.

The main reason for this approach (manually and atomically applying the new Ansible changes) is because our live Windows servers were configured long time ago (before we had Ansible), and I was concerned about Ansible conflicting with anything configured there.

I realize that this approach is not desirable at all, as we might end up in a situation like this one. Recently, instead of manually and atomically applying the Ansible changes, I tried to run all the Ansible tasks, everything succeeded, but the upgrade for Git brought our regression. That's exactly the reason you saw Git URL being `Git v2.20` in master, and not configured on our system until recently.

Right now, since our Ansible scripts are matured enough, it would take me no time to rebuild those nodes from clean ACC VMs, and make sure Ansible has no troubles running all the setup scripts on every check-in into master.

The Ansible scripts, once they configured a clean machine, they are meant to be idempotent, and on every subsequent run, they shouldn't break anything.

Therefore, to fully guard ourselves of these regressions due to the setup scripts (Ansible, or any other plain PowerShell / Bash scripts), in my opinion, we should include these, via Bors, on every code check-in:
1. Actually run the setup scripts (Ansible / PowerShell / Bash) to prepare an environment on clean ACC VMs. This means:
    * Spawn 3x clean ACC VMs with all our supported platforms: Ubuntu 16.04, Ubuntu 18.04 and Windows Server 2016.
    * Configure them, and see if the scripts finish successfully (this will catch regressions for the scripts)
    * Validate that the freshly configured environments are fine.
2. **After** every code check-in into master, run Ansible against the live CI servers, to make sure they stay updated every time we update our code.

I think that `#1` from above might be an overkill since it has these caveats:
* It depends on oe-engine and Azure to successfully deploy new VMs, and sometimes you might get transient errors for deployments (I noticed them from time to time on our oe-engine jobs);
* You have to wait for 3x ACC VMs to deploy, and this will slow down Bors.

The `#2` from above is all good! The only caveat here would be that regressions are caught **after** code merges. But if the patches are manually tested by the submitter, and carefully reviewed (I can help with this as I have a good understanding about the CI infrastructure), everything should be all right.

@CodeMonkeyLeet @anitagov @johnkord 

What do you think about these two options ?

> 
> > Given the ideal that we would like the developer set up scripts, our CI/CD environment and the ACC dev environment to be set up in consistent ways, I would prefer to reduce the number of explicit dependencies that need to be installed separately. i.e. using a standalone OpenSSL install for Windows in CD also means adding that to manual instructions for maintainers & ISV devs, otherwise, we would not be testing the configuration we assert we support.
> 
> There is an outstanding request for a power-shell script to install the tools on Windows as developers cannot run Ansible scripts on Windows.

This is work in progress from our side at PR https://github.com/microsoft/openenclave/pull/1934. Basically, for the Windows environment setup, we'll be having a standalone PowerShell script.

All the current Ansible Windows setup tasks will be converted into this new standalone PowerShell script, and our Ansible scripts will invoke the PowerShell script. This way, any Windows dev can configure an ACC VM by simply executing the PowerShell script, and our Ansible scripts will continue working in the same way.

And moving forward, we will have to maintain only the PowerShell script for the ACC Windows environment setup.

**Comment by CodeMonkeyLeet** [comment]
@ionutbalutoiu Regarding the listed options:
* Is `#2` expected to be a manual process? The potential latency between when a configuration change is deployed and when it starts getting applied to subsequent PRs is a concern, but requiring that someone manually remember to apply the new configuration is the part I would be most concerned about trying to eliminate from the process.
* I think the quality gating behavior of `#1` is what we should aim for, and we should consider engineering mitigations for the downsides.
    * For example, we should be able to determine the set of collateral that constitutes a deployment environment configuration change (i.e. jenkins & ansible files, deployment scripts) and only run the redeployment configuration when those are detected.
    * Note that once the configuration change is in the build, we would also want to make sure that the all CI/CD servers have the new configuration for subsequent runs, which is more complicated in this approach. It would need to:
        1. Deploy the 3 configurations, run them, check for failures
        1. On failure, discard/rollback the failed configurations.
        1. On success, apply the updated configuration to existing servers before allowing the next PRs to run on the new configuration.
    * As an intermediate step, it may be worth spinning off a different job configuration entirely to encapsulate this bit of engineering.

**Comment by ionutbalutoiu** [comment]
@CodeMonkeyLeet 

> @ionutbalutoiu Regarding the listed options:
> 
> * Is `#2` expected to be a manual process? The potential latency between when a configuration change is deployed and when it starts getting applied to subsequent PRs is a concern, but requiring that someone manually remember to apply the new configuration is the part I would be most concerned about trying to eliminate from the process.

Both `#1` and `#2` are expected to be fully automated.

The way I see `#2` being properly setup is:
* Configure a Jenkins job that will apply the Ansible scripts from the Open Enclave repositorygiven a SHA1 commit ID of the Open Enclave repository. Thus, when the job clones the repository, we know for sure that the revision 

> * I think the quality gating behavior of `#1` is what we should aim for, and we should consider engineering mitigations for the downsides.
>   
>   * For example, we should be able to determine the set of collateral that constitutes a deployment environment configuration change (i.e. jenkins & ansible files, deployment scripts) and only run the redeployment configuration when those are detected.
>   * Note that once the configuration change is in the build, we would also want to make sure that the all CI/CD servers have the new configuration for subsequent runs, which is more complicated in this approach. It would need to:
>     
>     1. Deploy the 3 configurations, run them, check for failures
>     2. On failure, discard/rollback the failed configurations.
>     3. On success, apply the updated configuration to existing servers before allowing the next PRs to run on the new configuration.
>   * As an intermediate step, it may be worth spinning off a different job configuration entirely to encapsulate this bit of engineering.

**Comment by ionutbalutoiu** [comment]
> @ionutbalutoiu Regarding the listed options:
> 
> * Is `#2` expected to be a manual process? The potential latency between when a configuration change is deployed and when it starts getting applied to subsequent PRs is a concern, but requiring that someone manually remember to apply the new configuration is the part I would be most concerned about trying to eliminate from the process.

Both `#1` and `#2` are expected to be fully automated.

The way I see `#2` being properly setup is:
* Configure a Jenkins job that will apply the Ansible scripts from the Open Enclave repository.
* Make sure that the Jenkins job is automatically triggered on every push event to the Open Enclave master branch. We will have Jenkins listen for GitHub webhooks triggered on push events.
* At this point, on every check-in, the job is automatically triggered and it applies the latest Ansible scripts.

> The potential latency between when a configuration change is deployed and when it starts getting applied to subsequent PRs is a concern

All our code check-ins go through [Bors staging job](https://oe-jenkins.eastus.cloudapp.azure.com/job/Bors/job/staging). This job is the result for the `bors r+` command, and its success causes a check-in or not. The `bors r+` requests are queued in the Jenkins CI, to avoid merge conflicts if two persons issue the command at the same time.

So, given the fact that the [Bors staging job](https://oe-jenkins.eastus.cloudapp.azure.com/job/Bors/job/staging) executes requests sequentially, and it takes on average about `30 mins` to run the full testing pipeline, my Jenkins job described above has no problems applying the Ansible changes in the meantime.

However, let's say that, somehow, the `Apply Ansible` job takes longer than 30 minutes, thus Bors having enough time to do another check-in, we would still be safe. I will make sure that if a job to apply the Ansible changes is running, a new one triggered by GitHub webhooks will be queued, and executed **after** the current one is finished.

Keep in mind that the option `#2` is not that difficult to setup (as oposed to `#1`), so it should guard us of any regressions until we get `#1` fully setup.

@CodeMonkeyLeet Does this sound good to you ?

> * I think the quality gating behavior of `#1` is what we should aim for, and we should consider engineering mitigations for the downsides.
>   
>   * For example, we should be able to determine the set of collateral that constitutes a deployment environment configuration change (i.e. jenkins & ansible files, deployment scripts) and only run the redeployment configuration when those are detected.
>   * Note that once the configuration change is in the build, we would also want to make sure that the all CI/CD servers have the new configuration for subsequent runs, which is more complicated in this approach. It would need to:
>     
>     1. Deploy the 3 configurations, run them, check for failures
>     2. On failure, discard/rollback the failed configurations.
>     3. On success, apply the updated configuration to existing servers before allowing the next PRs to run on the new configuration.

This would be the goal with option `#1`.

>   * As an intermediate step, it may be worth spinning off a different job configuration entirely to encapsulate this bit of engineering.

We might want to go with this approach at the beginning. Once we notice that this separate job behaves as expected, we might want to add it to our Bors testing pipeline, and run it on every new PR.

@CodeMonkeyLeet  What do you think ?

**Comment by anitagov** [comment]
@ionutbalutoiu - Thanks for listing out all the possibilities.
I agree with @CodeMonkeyLeet for the most part.
#1 is definitely overkill due to deploying clean ACC VMs but should also be tested periodically. We lose the capability of running scripts on existing ACC VMs or machines as that is usually the usage model. #2 will address this.
**To summarize**: My suggestion is to have #2 as the default method whenever CI is triggered and #1 be run whenever there is a change to the ansible/deployment scripts, tool upgrades, etc. Also believe that a new set of ACC VMs should be deployed in parallel to existing mainline CI and the switch made only after it is proven to be clean and stable.

Also wanted to confirm that we are still talking only about Windows here.

**Comment by ionutbalutoiu** [comment]
> @ionutbalutoiu - Thanks for listing out all the possibilities.
> I agree with @CodeMonkeyLeet for the most part.
> #1 is definitely overkill due to deploying clean ACC VMs but should also be tested periodically. We lose the capability of running scripts on existing ACC VMs or machines as that is usually the usage model. #2 will address this.
> **To summarize**: My suggestion is to have #2 as the default method whenever CI is triggered and #1 be run whenever there is a change to the ansible/deployment scripts, tool upgrades, etc. Also believe that a new set of ACC VMs should be deployed in parallel to existing mainline CI and the switch made only after it is proven to be clean and stable.
> 
> Also wanted to confirm that we are still talking only about Windows here.

@anitagov yes, we are still talking about Windows, but I think it's worth considering the other supported platforms as well.

I like your suggestion. Considering that, and our above conversation, I believe we should go with this approach:
1. When there's a check-in to the OE master branch that modifies the Ansible / deployment scripts, we should deploy a new set of ACC VMs in parallel to the existing mainline CI, and check if the update is clean and stable. The new set of ACC VMs should consist of whatever platforms are supported by OE. At this moment, the supported platforms are:
      * Ubuntu 16.04
      * Ubuntu 18.04
      * Windows Server 2016
2. If everything succeeded, and the update proved to be stable, we should do the following:
   * Apply the latest Ansible scripts to **all** our mainline CI machines (ACC Ubuntu 16.04, ACC Ubuntu 18.04, and ACC Windows Server 2016)
   * Update the latest Docker images using the updated Ansible / deployment scripts
   * Update the cached VHDs used for libcxx nightly job.

    Basically, we make sure that our live testing infrastructure is being updated.

**Comment by CodeMonkeyLeet** [comment]
@ionutbalutoiu A couple of follow-ups:

- For `#2`, I'm not clear on the robustness of:

    > So, given the fact that the Bors staging job executes requests sequentially, and it takes on average about 30 mins to run the full testing pipeline, my Jenkins job described above has no problems applying the Ansible changes in the meantime.

    It sounds like you're planning to run the Ansible job while the staging job is still running instead of after it has completed and is merged back to master? Regardless of how long the Ansible job you're proposing takes, it sounds like it would be necessary to complete the merge, block all other jobs (possibly across all pipelines, not just the Bors one), apply the changes to all CI servers, then drain the queue of CI/bors requests.

-  My understanding is that `#2` is only being proposed as a temporary until we can implement `#1`, is that correct? i.e. based on:

    > Keep in mind that the option #2 is not that difficult to setup (as oposed to #1), so it should guard us of any regressions until we get #1 fully setup.

    If so, that's probably okay. I do not agree with Anita that they should be run concurrently; once we can reliably identify changes to the git repo that result in CI environment changes for `#1` (i.e. playbooks & scripts), there's no point in a blanket job `#2` that always runs after every merge that introduces a non-zero overhead.

- Regarding your proposal, I agree with that it should apply to all configurations we currently support, not just Windows, and that the update requires all of the actions you have listed:

    > - Apply the latest Ansible scripts to all our mainline CI machines (ACC Ubuntu 16.04, ACC Ubuntu 18.04, and ACC Windows Server 2016)
    > - Update the latest Docker images using the updated Ansible / deployment scripts
    > - Update the cached VHDs used for libcxx nightly job.

    What I'm unclear on is whether you're suggesting this as a modification of '#2', or if you're referring to `#1` here:

    > **When there's a check-in to the OE master branch** that modifies the Ansible / deployment scripts, we should deploy a new set of ACC VMs in parallel to the existing mainline CI, and check if the update is clean and stable.

    If you mean this happens before we allow the change to merge (`#1`) then yes, I'm in agreement. If you mean this is still a separate job that runs after code has merged (`#2`) then I'd like a little more clarification; it's not clear to me what the value of running this as a separate job is if it needs to analyze a previous changeset to determine if it involved ansible/script changes.

**Comment by ionutbalutoiu** [comment]
@CodeMonkeyLeet see my responses below:

> @ionutbalutoiu A couple of follow-ups:
> 
> * For `#2`, I'm not clear on the robustness of:
>   > So, given the fact that the Bors staging job executes requests sequentially, and it takes on average about 30 mins to run the full testing pipeline, my Jenkins job described above has no problems applying the Ansible changes in the meantime.
>   
>   
>   It sounds like you're planning to run the Ansible job while the staging job is still running instead of after it has completed and is merged back to master?

No, the job is meant to run after the staging job is completed, and the Ansible code updates are checked-in to the master. Sorry if my comment was confusing here.

I just wanted to point out that we will make sure that the jobs will be properly  queued even if multiple `bors r+` are triggered at the same time.

>   Regardless of how long the Ansible job you're proposing takes, it sounds like it would be necessary to complete the merge, block all other jobs (possibly across all pipelines, not just the Bors one), apply the changes to all CI servers, then drain the queue of CI/bors requests.

Yep, that's precisely the plan with option `#2`!

> * My understanding is that `#2` is only being proposed as a temporary until we can implement `#1`, is that correct? i.e. based on:
>   > Keep in mind that the option #2 is not that difficult to setup (as oposed to #1), so it should guard us of any regressions until we get #1 fully setup.
>   
>   
>   If so, that's probably okay. I do not agree with Anita that they should be run concurrently; once we can reliably identify changes to the git repo that result in CI environment changes for `#1` (i.e. playbooks & scripts), there's no point in a blanket job `#2` that always runs after every merge that introduces a non-zero overhead.

I think that @anitagov recommended an improvement for the `#2`. After a check-in to the master branch, instead of **directly** applying the new Ansible changes to the live systems, we should apply them to some temporary new clean ACC machines **first**, and validate that the changes are stable. And **only** when the changes are stable, we apply them directly to the live systems.

I think that she is concerned that we might merge some Ansible changes, and at the review process we don't see a change that will result into a regression which will breaks the CI. But, given my understanding about the Jenkins machines, I can help with meaningful reviews for the PRs affecting our Jenkins testing infrastructure, and we shouldn't ultimately complicate the option `#2`.

Moreover, this suggested improvement sounds pretty much like the workflow for `#1`, implemented as an initial stage into `#2`. Maybe, after we finish `#1`, it's a good idea to **temporarily** implement it like this, until it proves reliable.

Once we see it reliable enough, we can transition `#1` as an extra check for Bors, and run it on every PR to the Ansible / setup scripts changes, **before** code is merged into master. I believe, this is the ultimate goal with `#1`.

Furthermore, even after reliably implementing `#1` into Bors, I still believe that we would need `#2` to apply the changes to our static Jenkins CI machines **after** the Ansible / setup scripts changes get merged into master. 

Otherwise, we need extra logic to the Jenkinsfile to detect `bors r+` vs a simple `bors try`. We want our live static systems to be updated after a successful `bors r+` testing pipeline (because this results into a check-in). But, a successful `bors try` pipeline shouldn't apply any changes to the live systems, because the PR might still be work in progress.

Thus, I think it's easier and cleaner to ultimately have option `#2`, together with option `#1` implemented for Bors.

> * Regarding your proposal, I agree with that it should apply to all configurations we currently support, not just Windows, and that the update requires all of the actions you have listed:
>   > * Apply the latest Ansible scripts to all our mainline CI machines (ACC Ubuntu 16.04, ACC Ubuntu 18.04, and ACC Windows Server 2016)
>   > * Update the latest Docker images using the updated Ansible / deployment scripts
>   > * Update the cached VHDs used for libcxx nightly job.
>   
>   
>   What I'm unclear on is whether you're suggesting this as a modification of '#2', or if you're referring to `#1` here:
>   > **When there's a check-in to the OE master branch** that modifies the Ansible / deployment scripts, we should deploy a new set of ACC VMs in parallel to the existing mainline CI, and check if the update is clean and stable.
>   
>   
>   If you mean this happens before we allow the change to merge (`#1`) then yes, I'm in agreement. If you mean this is still a separate job that runs after code has merged (`#2`) then I'd like a little more clarification; it's not clear to me what the value of running this as a separate job is if it needs to analyze a previous changeset to determine if it involved ansible/script changes.

Yep, indeed this is a modification for `#2`, suggested as an improvement by Anita.

But, as described above, it's exactly the workflow for `#1`, which will ultimately be an extra check for Bors, and it will run **before** the code merges.

**Comment by ionutbalutoiu** [comment]
To summarize our discussion from here, I've created the issue https://github.com/openenclave/openenclave/issues/1992.

**Comment by CodeMonkeyLeet** [comment]
@ionutbalutoiu continuing the discussion here and keeping #1992 as the clean summary of the outcome of this.

To me the question of whether having `#2` also check for Ansible/script changes makes sense or not depends on whether the Jenkins job can determine if it is performing a Bors stage or try request. Based on:

> Otherwise, we need extra logic to the Jenkinsfile to detect `bors r+` vs a simple `bors try`. We want our live static systems to be updated after a successful `bors r+` testing pipeline (because this results into a check-in). But, a successful `bors try` pipeline shouldn't apply any changes to the live systems, because the PR might still be work in progress.

Is doing that detection a problem? I had assumed that bors could specify it as part of kicking off the Jenkins run, but I looked at the bors configuration and it looks very simplistic. Assuming bors cannot parameterize its invocation of the Jenkin's job to specify staging/trying, is it possible to just have the Jenkins workflow detect whether the invocation is happening on the _staging_ or _trying_ branches? Since we require all merges to master to go through bors and its _staging_ branch, it should be sufficient to determine if the job should attempt to update the configuration of the CI/CD pipeline on a successful pass of the test.

If it is possible, the workflow I'm thinking of would look like:

1. Bors kicks off a Jenkins run on either _staging_ or _trying_ branches.
1. Jenkins job checks if Ansible playbooks or deployment scripts were modified in the PR/branch being tested.
    1. If so, setup the 3 new configurations using the PR/branch on clean ACC VMs and run the rest of the workflow on them. Otherwise, use the existing devices.
1. Jenkins job builds OE and runs all required ctests on all configurations.
1. If build and tests succeed, check if whether it was running new configurations **and** whether  success is on the bors _staging_ branch.
    1. If so, update all CI VMs to use the new configuration.
        1. As an optimization, we can take the newly setup servers used in the test and put them into rotation, while taking the old ones out, then provision enough new additional servers to make up the difference in numbers.
    1. If not the new VMs can be recycled. 
1. Bors gets a success Jenkins run result back, and merges the PR into master if it was a staging run.

Even in this case, I'm unclear on the synchronization of tasks running in bors:

- We have mentioned that bors jobs must happen sequentially, but is a bors job allowed to run concurrently with a different pipeline job that uses the same VM pool? (e.g. a bors job and a sandbox job running concurrently). 
    - If so, the VM configuration update task needs to wait for other jobs to finish and prevent new (non-bors) jobs from pre-empting it on those VMs. Is this feasible?
    - e.g. if configuration changes to existing VMs can be enqueued to run after existing jobs complete, and the jenkins job only finishes once all configuration tasks have reported completion, that would be sufficient.

In the case where we want to take 4.i. above and make it a separate task as in `#2`:

1. Bors kicks off a Jenkins run on either _staging_ or _trying_ branches.
1. Jenkins job does 2. & 3. from the previous case but skips 4.
1. Bors gets a success Jenkins run result back, and merges the PR into master if it was a staging run.
1. GitHub fires a webhook event on any merge to master that Jenkins uses:
    1. If this job did not filter on whether there were Ansible/script changes, then it would block the pipeline for for an additional duration of the deployment. 
    1. If this job did filter, it would mostly reuse the same code as 2.i. since the push webhook event provides the PR that was merged, correct? If so, we should just do that. 
1. Jenkins deploys the configuration update to all servers, blocking the bors and other pipelines until it is completed.
    1. The synchronization question for his job on different pipelines remains the same as in the `#1` case above.

In addition, I have some other questions specific to `#2`:

- Isn't there a race condition between Jenkins receiving the GitHub webhook event and another bors r+ job that was already in queue being allowed to run? Is there some additional mechanism you're proposing to prevent this?
- The replacement/rollback strategy is a little bit less clear to me if it spans two jenkins jobs instead of one. 
    - It sounds like under this scheme, we would update existing machines, which could carry forwards previous state that make it slightly different from the test configuration. If there is a partial failure (e.g. one of the servers failed to update) I don't know what the plan is to restore the state of the already updated servers.
    - Assuming we're using a replacement strategy, then it's less efficient that the single job case where it could potentially just use the new VMs it set up for test already. Or are you suggesting some mechanism to maintain state between two jobs? If so, what's the cleanup strategy if the second job fails or never runs?

**Comment by ionutbalutoiu** [comment]
@CodeMonkeyLeet 

> @ionutbalutoiu continuing the discussion here and keeping #1992 as the clean summary of the outcome of this.
> 
> To me the question of whether having `#2` also check for Ansible/script changes makes sense or not depends on whether the Jenkins job can determine if it is performing a Bors stage or try request. Based on:
> 
> > Otherwise, we need extra logic to the Jenkinsfile to detect `bors r+` vs a simple `bors try`. We want our live static systems to be updated after a successful `bors r+` testing pipeline (because this results into a check-in). But, a successful `bors try` pipeline shouldn't apply any changes to the live systems, because the PR might still be work in progress.
> 
> Is doing that detection a problem? I had assumed that bors could specify it as part of kicking off the Jenkins run, but I looked at the bors configuration and it looks very simplistic. Assuming bors cannot parameterize its invocation of the Jenkin's job to specify staging/trying, is it possible to just have the Jenkins workflow detect whether the invocation is happening on the _staging_ or _trying_ branches? Since we require all merges to master to go through bors and its _staging_ branch, it should be sufficient to determine if the job should attempt to update the configuration of the CI/CD pipeline on a successful pass of the test.

You are right. Detecting whether it's a `bors try` or a `bors r+` seems to be a trivial task since Jenkins sets an additional environment variable called `BRANCH_NAME` (documented [here](https://jenkins.io/doc/book/pipeline/multibranch/#additional-environment-variables)).

We would have to check against this. If it is `staging`, and the job is successful, we know that we have stable changes resulting in a check-in, and we can apply everything on the static systems already (assuming that we detected deployment scripts changes, as mentioned below).

Thus, `#2` will be redundant and replaced once `#1` is fully setup.

> 
> If it is possible, the workflow I'm thinking of would look like:
> 
> 1. Bors kicks off a Jenkins run on either _staging_ or _trying_ branches.
> 2. Jenkins job checks if Ansible playbooks or deployment scripts were modified in the PR/branch being tested.
>    
>    1. If so, setup the 3 new configurations using the PR/branch on clean ACC VMs and run the rest of the workflow on them. Otherwise, use the existing devices.

I realized that we have some containerized Docker testing stages, and the Docker images are generated with Ansible, as well. So we need to build new Docker images specific to the changes in the pull request, and use these for testing.

There's the tag `latest` that it's used all the time, right now. The `latest` tag will be updated before merging into master if we have stable changes. This action will done together with applying changes to the static systems (considering that all the conditions are met). 

However, there's one small caveat that I see here. If we run the rest of the testing workflow only on the new 3 configurations, the full testing pipeline will probably take more time.

Currently the Jenkins CI has the following static systems:
* 3x ACC Ubuntu 16.04 systems
* 3x ACC Ubuntu 18.04 systems
* 2x ACC Windows Server 2016 systems
* 2x non-ACC Ubuntu 16.04 systems. These two systems are used to run the Docker containerized nonSGX testing stages for OE simulation + ARM cross platform build validation:
https://github.com/openenclave/openenclave/blob/9ca242255b19a1b7acece64f59bb5e1479907dba/.jenkins/Jenkinsfile#L217-L228

If we detect deployment scripts changes, and we configure 3x new ACC VMs, the rest of the testing pipeline will run without 2x ACC 16.04, 2x ACC 18.04, and 1x ACC Win2016 compared to the usual Bors jobs against the static systems.

The OE simulation stages + ARM compilation validation stages can be safely scheduled on the existing static non-ACC VMs from Jenkins, but using the Docker images generated on top of the pull request. So, these stages won't slow down anything.

An usual Bors job takes about `21 minutes` on the static systems, right now. If the job execution times won't be slowed down significantly (let's say that we still stay under 45 minutes ?), I believe that we are fine. We'll see how it goes.

What do you think ? What executing time would you still find reasonable ?

There's the option to have more than a single worker / ACC VM. But in the past, I think that we didn't see impressive improvements for the execution time. We can try and experiment with this again, if the Bors job execution time gets unreasonably slowed down.

> 3. Jenkins job builds OE and runs all required ctests on all configurations.
> 4. If build and tests succeed, check if whether it was running new configurations **and** whether  success is on the bors _staging_ branch.
>    
>    1. If so, update all CI VMs to use the new configuration.
>       
>       1. As an optimization, we can take the newly setup servers used in the test and put them into rotation, while taking the old ones out, then provision enough new additional servers to make up the difference in numbers.
>    2. If not the new VMs can be recycled.
> 5. Bors gets a success Jenkins run result back, and merges the PR into master if it was a staging run.
> 

Besides my above comment, I like this workflow!

It's a robust implementation of the proposed `#1`, and I agree with it.

> Even in this case, I'm unclear on the synchronization of tasks running in bors:
> 
> * We have mentioned that bors jobs must happen sequentially, but is a bors job allowed to run concurrently with a different pipeline job that uses the same VM pool? (e.g. a bors job and a sandbox job running concurrently).

Yes, the Bors job is allowed to run concurrently with a different pipeline jobs that use the same VM pool. Pipelines testing stages are scheduled on free workers from the VMs pool, and there's a global queue that Jenkins uses in case all the workers are occupied.

>   
>   * If so, the VM configuration update task needs to wait for other jobs to finish and prevent new (non-bors) jobs from pre-empting it on those VMs. Is this feasible?
>   * e.g. if configuration changes to existing VMs can be enqueued to run after existing jobs complete, and the jenkins job only finishes once all configuration tasks have reported completion, that would be sufficient.

The configuration changes to the existing VMs can be enqueued. And I agree that this would be sufficient, as long as the existing jobs complete, and then the tasks to update the VM configuration start. 

> 
> In the case where we want to take 4.i. above and make it a separate task as in `#2`:
> 
> 1. Bors kicks off a Jenkins run on either _staging_ or _trying_ branches.
> 2. Jenkins job does 2. & 3. from the previous case but skips 4.
> 3. Bors gets a success Jenkins run result back, and merges the PR into master if it was a staging run.
> 4. GitHub fires a webhook event on any merge to master that Jenkins uses:
>    
>    1. If this job did not filter on whether there were Ansible/script changes, then it would block the pipeline for for an additional duration of the deployment.
>    2. If this job did filter, it would mostly reuse the same code as 2.i. since the push webhook event provides the PR that was merged, correct? If so, we should just do that.

Yep, I think this job should filter, and it's a reuse of the code for 2.i.

I believe that the GitHub push webhook event should give us enough information for the PR that just got merged. So, we should be good here.

> 5. Jenkins deploys the configuration update to all servers, blocking the bors and other pipelines until it is completed.
>    
>    1. The synchronization question for his job on different pipelines remains the same as in the `#1` case above.

We are aware of this. The whole idea is to enqueue the configuration changes job.

We have a single worker / ACC VM, so as long as you queue the VMs configuration update job on every Jenkins CI machine, we are good.

> 
> In addition, I have some other questions specific to `#2`:
> 
> * Isn't there a race condition between Jenkins receiving the GitHub webhook event and another bors r+ job that was already in queue being allowed to run? Is there some additional mechanism you're proposing to prevent this?

I thought about this one, and indeed it applies only to `#2`. So, there is a point at the beginning of every testing stage where the SCM code is checked out. Instead of checking out latest master here, I wanted to have a check-out exactly to the latest commit when the GitHub push event was received.

I believe that the GitHub webhook push event should give us information about this. If that is not the case, I would have, at the beginning of the pipeline, an initial step that saves the latest commit ID, and use this whenever we checkout the code.

I would also not allow the Jenkins job for `#2` to have multiple concurrent runs. 

> * The replacement/rollback strategy is a little bit less clear to me if it spans two jenkins jobs instead of one.
>   
>   * It sounds like under this scheme, we would update existing machines, which could carry forwards previous state that make it slightly different from the test configuration. If there is a partial failure (e.g. one of the servers failed to update) I don't know what the plan is to restore the state of the already updated servers.

This is a very good point! We totally need a rollback mechanism for the Ansible update configuration job in case this fail, if we continue to go this route.

>   * Assuming we're using a replacement strategy, then it's less efficient that the single job case where it could potentially just use the new VMs it set up for test already. Or are you suggesting some mechanism to maintain state between two jobs? If so, what's the cleanup strategy if the second job fails or never runs?

I like the replacement strategy more, instead of updating the current state of the systems, which might lead to unexpected failures.

I think that it would be easier to simply replace the existing VMs in the Jenkins pool. So, considering that `#1` is properly implemented following your above workflow, I'm thinking something like:

1. Bors detected Ansible/scripts changes, tested them on new ACC VMs, and everything is green.
2. Set offline any VM in the current Jenkins VM pool.
   * Keep in mind that setting an agent offline in Jenkins will **not** stop any job already executing on them. Instead, it will **not** queue any other new task on it. This is what we want, because we want to deprovision the agents once they finish their current tasks.  
   * Join the newly created ACC VMs for testing to the existing Jenkins VM pool, making them static for future runs when no Ansible/scripts changes were detected.
3. The staging job can end successfully, and trigger a merge into master.

Some notes here:
* We probably want to enlarge the VM pool to reach the current Jenkins machines configuration (described above), or we could simply start to a number of testing ACC VMs that could simply fill the entire existing Jenkins VM pool
* We want to cleanup any VMs from the Jenkins VM pool set offline at step 2. I wouldn't do this from Bors, but instead do it with a separate job that gets triggered on every merge into master.

  Moreover, when setting up the VMs offline at step 2, we could add a new label called `to-be-deleted`, and this separate job will only try and deprovision the VMs without any workload on them and labelled `to-be-deleted`. If there's still existing workload present, it means that when they were set offline at step 2, there was something already executing on them, and we just wait for it to finish before deprovisioning.

What do you think about this plan ?
